### PR TITLE
add support for simple ${var} string interpolation expressions

### PR DESCRIPTION
**Pull request by quasilyte**
This syntax is preferred by some programmers over {$...}.

The main reasoning behind this change is to make more PHP
programs compile as KPHP programs.

A more compilcated ${<expr>} form is not implemented, so
this will result in a compile error: "${a[0]}".
We may add more complete support for this syntax later.

Note: <expr> inside ${<expr>} is subtle and should
be handled separately, if ever attempted to be implemented.
For example, "${'foo'}" is identical to "$foo".
We don't support ${'foo'} variable syntax yet (which is a
valid freestanding PHP variable syntax btw).

Signed-off-by: Iskander Sharipov <quasilyte@gmail.com>

### CODE REVIEW

**Comment by quasilyte** [comment]
Do not merge this until it's fully discussed whether this feature should be implemented.

**Comment by eugene536** [comment]
Well yes, but actually no

this piece of code compiles without any error
```
<?php

$a = 10;
${a} = 30;
var_dump("${a}");

function foo(${a}) {
    var_dump(${a});
}

foo(20);

/**@var ${uu} int*/
$uu = 20;

```

**Comment by quasilyte** [comment]
> this piece of code compiles without any error

Hmm. KPHP gives compile errors for it.

```
Compilation error [gen by /home/isharipov/CODE/kphp3/compiler/lexer.cpp at 294]
In stage = [Split file to tokens]:
  [file = /home/isharipov/CODE/php/hello.php]
  [function = unknown function:4]
  [line = 4]
//   4:${a} = 30;
${ is not supported by kPHP


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compilation error [gen by /home/isharipov/CODE/kphp3/compiler/lexer.cpp at 294]
In stage = [Split file to tokens]:
  [file = /home/isharipov/CODE/php/hello.php]
  [function = unknown function:4]
  [line = 4]
//   4:${a} = 30;
Can't parse


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compilation error [gen by /home/isharipov/CODE/kphp3/compiler/lexer.cpp at 1176]
In stage = [Split file to tokens]:
  [file = /home/isharipov/CODE/php/hello.php]
  [function = unknown function:4]
  [line = 4]
//   4:${a} = 30;
failed to parse
```

PHP also doesn't allow the `${x}` syntax in the param names. The main question for this PR is whether we want to support `${x}` syntax in strings. It's quite rare (one can always use `{$x}` if explicit delimiters are needed.

**Comment by quasilyte** [comment]
Ah. Now I see what you've meant. I'll fix that.

File: compiler/lexer.cpp
Comment (eugene536) [inline compiler/lexer.cpp:815]: it seems like a subtle bug here is hidden, maybe we unify these Lexers?
image

Comment (quasilyte) [reply compiler/lexer.cpp]: Fixed. Relevant tests are added.

**Review by eugene536 — APPROVED** [review]

**Comment by quasilyte** [comment]
Approve disappeared after rebase. I tested the branch, it gives no translation diff.

**Review by eugene536 — APPROVED** [review]

**Review by unserialize — APPROVED** [review]

**Pull request merged**
